### PR TITLE
Async borrowing (continuation of #449)

### PR DESCRIPTION
**Pull request by tanriol**
It seems to work as expected.

Todo:
- [x] Try to avoid mapping every error with `into_handler_error` before `?`

Feedback needed:
- [x] Could this allow for generic return types like `R: IntoResponse`? Yes.
- [x] Could this allow for handlers receiving either `&State` or `&mut State`? Not really.
- [x] Could this allow for returning the reponse without Ok-wrapping in the infallible case? Not really.

And "do we want that" for every question above :-)

### CODE REVIEW

File: gotham/src/router/builder/single.rs [hunk 1]
```diff
@@ -16,6 +16,24 @@ use crate::state::State;
...
+    type Res: IntoResponse + 'static;
+    type Fut: std::future::Future<Output = Result<Self::Res, HandlerError>> + Send + 'a;
+    fn call(self, arg: &'a mut State) -> Self::Fut;
+}
+
+impl<'a, D: 'a, R, F> AsyncHandlerFn<'a> for F
+where
+    F: FnOnce(&'a mut State) -> D,
+    R: IntoResponse + 'static,
+    D: std::future::Future<Output = Result<R, HandlerError>> + Send + 'a,
```
Comment (technic) [inline gotham/src/router/builder/single.rs:29]: The `D: 'a` requirement is duplicated in the line 25

Comment (tanriol) [reply gotham/src/router/builder/single.rs:29]: Removed it from line 25.

File: gotham/src/lib.rs [hunk 2]
```diff
@@ -3,8 +3,8 @@
 //! You can find out more about Gotham, including where to get help, at <https://gotham.rs>.
 //!
 //! We look forward to welcoming you into the Gotham community!
-#![doc(html_root_url = "https://docs.rs/gotham/0.5.0-rc.1")] // Update when changed in Cargo.toml
-#![warn(missing_docs, deprecated)]
+#![doc(html_root_url = "https://docs.rs/gotham/0.5.0-rc.1")]
+// Update when changed in Cargo.toml
```
Comment (alsuren) [inline gotham/src/lib.rs:7]: This was just me trying to remove the noise from my `cargo watch` output. Probably revert this and fix it in a separate PR.

File: gotham/src/router/builder/single.rs [hunk 3]
```diff
@@ -16,6 +16,24 @@ use crate::state::State;
...
+    type Res: IntoResponse + 'static;
+    type Fut: std::future::Future<Output = Result<Self::Res, HandlerError>> + Send + 'a;
+    fn call(self, arg: &'a mut State) -> Self::Fut;
+}
+
+impl<'a, D, R, F> AsyncHandlerFn<'a> for F
+where
+    F: FnOnce(&'a mut State) -> D,
+    R: IntoResponse + 'static,
+    D: std::future::Future<Output = Result<R, HandlerError>> + Send + 'a,
```
Comment (alsuren) [inline gotham/src/router/builder/single.rs:29]: this was only called D because the original gist was trying to return `impl Debug + 'a`. Probably want some better names for this.

File: gotham/src/service/trap.rs [hunk 4]
```diff
@@ -72,8 +72,8 @@ fn finalize_error_response(
         // API doesn't guarantee its presence (even though it always is).
         let err_description = err
             .source()
-            .map(Error::description)
-            .unwrap_or_else(|| err.description());
+            .map(Error::to_string)
+            .unwrap_or_else(|| err.to_string());
```
Comment (alsuren) [inline gotham/src/service/trap.rs:76]: Let's revert this too, and put it in its own PR.

File: gotham/src/router/builder/single.rs [hunk 5]
```diff
@@ -157,6 +175,31 @@ pub trait DefineSingleRoute {
...
+                    let result = fut.await;
+                    match result {
+                        Ok(data) => {
+                            let response = data.into_response(&state);
+                            Ok((state, response))
+                        }
+                        Err(err) => Err((state, err)),
+                    }
+                }
+                .boxed()
```
Comment (alsuren) [inline gotham/src/router/builder/single.rs:199]: is it possible to move this up into the AsyncHandlerFn as a `call_wrapped(State) -> Box<dyn Future<Output = Result<(State, _), (State, _)>>`? At the moment, there are a lot of complex moving parts that are spread all over the file.

File: gotham/src/lib.rs (hunk 2 above)
Comment (tanriol) [reply gotham/src/lib.rs:7]: Should I revert these or remove these changes when rebasing the patchset? I'm going to rebase my part anyway, but not sure whether I should touch your commit too or leave it as-is.

File: gotham/src/router/builder/single.rs (hunk 5 above)
Comment (tanriol) [reply gotham/src/router/builder/single.rs:199]: It may be possible to move this into a free function or another trait, but it explicitly cannot be moved into `AsyncHandlerFn` trait itself because it uses HRTB on that.

File: gotham/src/lib.rs (hunk 2 above)
Comment (alsuren) [reply gotham/src/lib.rs:7]: I'm happy for you to keep it as a separate commit or fixup my commit if you want. Up to you.

**Comment by tanriol** [comment]
I've tried making this work for an async function of `&State`. The problem is that for the future to be `Send` we'd need `&State` to be `Send`, which, according to the standard rules for shared references, requires `State` to be `Sync`.

**Comment by tanriol** [comment]
Looks like this would not work without `Result` in the infallible case because of ambiguous error type. The problem here is that there's the following impl
```rust
impl<T, E> IntoResponse for Result<T, E> where
    T: IntoResponse,
    E: IntoResponse, 
{ ... }
```
which means that returning `impl IntoResponse` and passing the errors with `?` will result in Rust being unable to infer the `E` type from the `ErrorTypeN: Into<E>` and `E: IntoResponse` bounds.

**Comment by technic** [comment]
As far as I understand rust uses `From` trait for converting between types when `?` operator is used (or `Into`). 
So if the want to go via intermediate type which implements `IntoResponse` there is no way for rust to deduce this type automatically.

**Comment by tanriol** [comment]
Pushed an updated version. I haven't looked into `IntoHandlerError` because @msrd0 said that may go in #438.

The main question, IMHO, is whether we want to keep both `to_async` and `to_async_borrowed` or just one of them?

**Comment by tanriol** [comment]
@technic 
> As far as I understand rust uses `From` trait for converting between types when `?` operator is used (or `Into`).
> So if the want to go via intermediate type which implements `IntoResponse` there is no way for rust to deduce this type automatically.

Not exactly. Rust can handle inference like "there's an Ok being returned, so this is a Result, and the only impl for a Result is an impl for `Result<_, HandlerError>`, so the error type is known to be a HandlerError". If the impl above were an `impl<T: IntoResponse> IntoResponse for Result<T, HandlerError>`, it would compile successfully.

**Comment by msrd0** [comment]
What is the status of this pull request? #438 has finally been merged, so this PR should no longer be blocked on it.

**Comment by tanriol** [comment]
Will rebase in a couple days (maybe faster if the rebase is easy). IIRC, there's the problem that two different variants allow either borrowing (and passing an async function) in async or passing a closure (but no borrowing), not both.

**Comment by alsuren** [comment]
I will admit that I've stepped away from both of these projects, but I occasionally get notifications from both projects, so I'm leaving this here in case it provides inspiration for anyone who comes along.

The people over at the goose project seem to have made some headway with closures. Last I saw, they ask the user to wrap the closure in an Arc and Box::pin() the resulting future themselves (the Arc is goose-specific, because of how the containing struct derives Clone. Gotham could probably use Box there too. See https://github.com/tag1consulting/goose/pull/120/files#diff-1cd6d8660ef657deb6ee2a73b4d592d9R1922-R1926 https://github.com/tag1consulting/goose/pull/120/files#diff-8484c6c853111ca2278cc1dc616df376R189-R195 and https://github.com/tag1consulting/goose/pull/126/files) but there might have been some ergonomics improvements since then that I've not been keeping on top of (I'm not sure whether this approach went anywhere, for example: https://github.com/tag1consulting/goose/pull/94#issuecomment-665741431 . I suspect not). 

I suspect that supporting closures is more interesting for goose than it is for gotham. If the ergonomics of some theoretical `.to_borrowing_closure()` are a bit less nice than the owning .to() and `.to_async()` variants then maybe that's fine. It might be an argument for keeping the old variants around though, until rustc is capable of letting you express the appropriate types and lifetimes for your function arguments directly.

**Review by sezna — APPROVED** [review]

**Review by msrd0 — APPROVED** [review]

**Comment by technic** [comment]
Great job! thanks!
